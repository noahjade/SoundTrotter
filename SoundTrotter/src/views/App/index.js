import React, { useEffect, useState } from "react";
import style from "./style.module.scss";
import {
  getCodeFromUrl,
  fetchAuthToken,
  fetchRefreshedToken,
  findExpiryTime,
} from "../../utils/spotifyService";

import SpotifyWebApi from "spotify-web-api-js";
// https://github.com/JMPerez/spotify-web-api-js

import Login from "../Login";
import Home from "../Home";

const STORAGE_ACCESS_TOKEN = "spotify_access_token";
const STORAGE_REFRESH_TOKEN = "spotify_refresh_token";
const STORAGE_TOKEN_EXPIRATION = "spotify_token_expiration";

const spotify = new SpotifyWebApi();

function App() {
  const [token, setToken] = useState("");

  const handleNewToken = (code) => {
    fetchAuthToken(code).then((tokenData) => {
      // Store the tokens and expiry time in local storage
      const accessToken = tokenData.access_token;
      const refreshToken = tokenData.refresh_token;
      const expiryTime = findExpiryTime(tokenData.expires_in);

      localStorage.setItem(STORAGE_ACCESS_TOKEN, accessToken);
      localStorage.setItem(STORAGE_REFRESH_TOKEN, refreshToken);
      localStorage.setItem(STORAGE_TOKEN_EXPIRATION, expiryTime);

      spotify.setAccessToken(accessToken);
      setToken(accessToken);

      window.location = "/";
    });
  };

  const handleRefreshToken = (refreshToken) => {
    fetchRefreshedToken(refreshToken).then((refreshData) => {
      if (refreshData) {
        // set the new access token
        const newToken = refreshData.access_token;
        const expiryTime = findExpiryTime(refreshData.expires_in);

        localStorage.setItem(STORAGE_ACCESS_TOKEN, newToken);
        localStorage.setItem(STORAGE_TOKEN_EXPIRATION, expiryTime);

        spotify.setAccessToken(newToken);
        setToken(newToken);
      } else {
        // If the token could not be refreshed, the user needs to log in again
        setToken("");
      }
    });
  };

  useEffect(() => {
    // See if there is an existing token in storage
    const existingToken = localStorage.getItem(STORAGE_ACCESS_TOKEN);
    if (existingToken) {
      // Check whether current token has expired
      const expiryTime = localStorage.getItem(STORAGE_TOKEN_EXPIRATION);
      if (!expiryTime || expiryTime <= Date.now()) {
        // Current token is expired
        // Try to get new one through refresh token
        const refreshToken = localStorage.getItem(STORAGE_REFRESH_TOKEN);
        handleRefreshToken(refreshToken);
      } else {
        // Localstorage token is not expired, so we can use it
        spotify.setAccessToken(existingToken);
        setToken(existingToken);
      }
    } else {
      // No localStorage token, try to get from url
      const urlData = getCodeFromUrl();

      if (urlData.code) {
        // New token from url
        handleNewToken(urlData.code);
      }
    }
  }, []);

  return (
    <div className={style.App}>
      {token ? (
        <Home spotify={spotify} />
      ) : (
        <div className={style.Wrapper}>
          {" "}
          <Login />{" "}
        </div>
      )}
    </div>
  );
}

export default App;
