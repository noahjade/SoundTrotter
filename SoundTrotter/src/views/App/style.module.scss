.App {
  text-align: center;
  height: 100%;
  width: 100%;
  font-size: 16px;
}

.Wrapper {
  height: 500px;
  width: 60%;
  background-color: #282828;
  top: 50%;
  left: 50%;
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  position: absolute;
  margin: auto;
}
