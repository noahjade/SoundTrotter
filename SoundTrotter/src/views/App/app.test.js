import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import App from ".";

const { act } = renderer;
jest.mock("../../utils/spotifyService");

const storeNewTokenData = () => {
  localStorage.setItem("spotify_access_token", "token");
  localStorage.setItem("spotify_token_expiration", Date.now());
  localStorage.setItem("spotify_refresh_token", "refresh-token");
};
const storeRefreshedTokenData = () => {
  localStorage.setItem("spotify_access_token", "new-token");
  localStorage.setItem("spotify_token_expiration", Date.now() + 60 * 60 * 1000);
  localStorage.setItem("spotify_refresh_token", "refresh-token");
};

afterEach(() => {
  // Reset storage to remove token data
  localStorage.clear();

  // Reset jest function mocks
  jest.clearAllMocks();
});

test("Correctly renders default login screen", () => {
  // Shallow render used so the useEffect block won't run
  // If this runs, the component will get a token from mockSpotifyService
  // This authenticates the user, so the login screen won't show
  const defaultAppScreen = shallow(<App />);
  expect(defaultAppScreen).toMatchSnapshot();
});

test("It tries to get the spotify code from the url on each render", async () => {
  const spotifyService = require("../../utils/spotifyService");
  const urlCodeSpy = jest.spyOn(spotifyService, "getCodeFromUrl");

  await act(async () => {
    renderer.create(<App />);
  });
  expect(urlCodeSpy).toBeCalledTimes(1);
});

test("It tries to get access token data using the url code", async () => {
  const spotifyService = require("../../utils/spotifyService");
  const fetchAuthTokenSpy = jest.spyOn(spotifyService, "fetchAuthToken");

  await act(async () => {
    renderer.create(<App />);
  });
  expect(fetchAuthTokenSpy).toBeCalledTimes(1);
});

test("It uses the findExpiryTime utility for a new token", async () => {
  const spotifyService = require("../../utils/spotifyService");
  const findExpiryTimeSpy = jest.spyOn(spotifyService, "findExpiryTime");

  await act(async () => {
    renderer.create(<App />);
  });
  expect(findExpiryTimeSpy).toBeCalledTimes(1);
});

test("It stores new token items in localStorage", async () => {
  await act(async () => {
    renderer.create(<App />);
  });

  expect(Object.keys(localStorage.__STORE__).length).toBe(3);
  expect(localStorage.__STORE__["spotify_access_token"]).toBeDefined();
  expect(localStorage.__STORE__["spotify_refresh_token"]).toBeDefined();
  expect(localStorage.__STORE__["spotify_token_expiration"]).toBeDefined();
});

test("It tries to get a refreshed token when the current token has expired", async () => {
  const spotifyService = require("../../utils/spotifyService");
  const fetchRefreshedTokenSpy = jest.spyOn(
    spotifyService,
    "fetchRefreshedToken"
  );

  storeNewTokenData();

  await act(async () => {
    renderer.create(<App />);
  });

  expect(fetchRefreshedTokenSpy).toBeCalledTimes(1);
});

test("It stores the refreshed token information", async () => {
  storeNewTokenData();

  await act(async () => {
    renderer.create(<App />);
  });

  // new-token is only given by the mock freshToken function
  expect(localStorage.__STORE__["spotify_access_token"]).toBe("new-token");
});

test("It doesnt call any token methods when there is a good token in storage", async () => {
  const spotifyService = require("../../utils/spotifyService");
  const findExpiryTimeSpy = jest.spyOn(spotifyService, "findExpiryTime");
  const fetchRefreshedTokenSpy = jest.spyOn(
    spotifyService,
    "fetchRefreshedToken"
  );
  const fetchAuthTokenSpy = jest.spyOn(spotifyService, "fetchAuthToken");

  storeRefreshedTokenData();

  await act(async () => {
    renderer.create(<App />);
  });
  expect(findExpiryTimeSpy).toBeCalledTimes(0);
  expect(fetchRefreshedTokenSpy).toBeCalledTimes(0);
  expect(fetchAuthTokenSpy).toBeCalledTimes(0);
});

test("It renders the home page for an authorised user", async () => {
  storeRefreshedTokenData();

  // DOM render used so the mockSpotifyService can be used
  let appHomePage;
  await act(async () => {
    appHomePage = renderer.create(<App />);
  });
  expect(appHomePage).toMatchSnapshot();
});
