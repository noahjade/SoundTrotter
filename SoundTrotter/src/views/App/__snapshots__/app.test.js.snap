// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly renders default login screen 1`] = `
<div
  className="App"
>
  <div
    className="Wrapper"
  >
     
    <Login />
     
  </div>
</div>
`;

exports[`It renders the home page for an authorised user 1`] = `
<div
  className="App"
>
  <div
    className="home"
  >
    <div
      className="topBar"
    >
      <div
        className="topBarLeft"
      >
        <img
          alt="Logo"
          src="spotify-icon-green.png"
        />
        <h3
          className="spaceItem"
        >
          SOUND TROTTER
        </h3>
      </div>
      <div
        className="topBarRight"
      >
        <h3
          className="spaceItem"
        >
          Welcome, 
          
        </h3>
        <img
          alt="help"
          className="spaceItem"
          onClick={[Function]}
          src="helpButton.png"
        />
      </div>
    </div>
    <div
      className="helpContainer"
      showhelp={0}
    >
      <div
        className="helpBox"
      >
        <div
          className="rightAlignRow"
        >
          <div
            className="ExitButton"
            onClick={[Function]}
          />
        </div>
        <div
          className="titleArea"
        >
          Welcome to Sound Trotter!
        </div>
        <div
          className="contentArea"
        >
          <p>
            This application allows you to analyse your recent spotify listening to find out which countries the songs you have been listening to are from. Simply wait for the data to load in, and the interactive map will light up green for every country you have covered.
          </p>
          <p>
            The tabs on the right side of the screen allow you to analyse further. Click on them to explore! Select on the map to view specific information about specific countries.
          </p>
        </div>
      </div>
    </div>
    <div
      className="centerContent"
    >
      <div
        className="mapArea"
      />
      <div
        className="sideBar"
      >
        <div
          className="MainWrapper"
        >
          <div
            className="TabWrapper"
            hidemain={1}
          >
            <div
              className="Tab tabOne"
              onClick={[Function]}
              tabNumber={0}
            >
              <img
                alt="User Stats"
                src="user.png"
              />
            </div>
            <div
              className="Tab tabTwo"
              onClick={[Function]}
              tabNumber={0}
            >
              <img
                alt="Country Artists"
                src="flag.png"
              />
            </div>
            <div
              className="Tab tabThree"
              onClick={[Function]}
              tabNumber={0}
            >
              <img
                alt="Country Charts"
                src="headphones.png"
              />
            </div>
            <div
              className="Tab tabFour"
              onClick={[Function]}
              tabNumber={0}
            >
              <img
                alt="Credits"
                src="heart.png"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="bottomBar"
    >
      <section
        className="hidden"
      >
        <div
          className="track"
        >
          <div
            className="image"
          >
            <img />
          </div>
          <div
            className="info"
          >
            <span
              className="info__title"
            />
            <span
              className="info__artists"
            />
          </div>
          <div
            className="track__controls"
          >
            <button
              className="btn"
              id="previous-btn"
              onClick={[Function]}
            >
              <svg>
                prev.svg
              </svg>
            </button>
            <button
              className="btn"
              disabled={false}
              id="play-btn"
              onClick={[Function]}
            >
              <svg>
                play.svg
              </svg>
            </button>
            <button
              className="btn"
              id="next-btn"
              onClick={[Function]}
            >
              <svg>
                next.svg
              </svg>
            </button>
          </div>
        </div>
      </section>
    </div>
  </div>
</div>
`;
